r/tests/testthat/test-dataset-csv.R

```diff
@@ -299,3 +299,21 @@ test_that("open_dataset() deals with BOMs (byte-order-marks) correctly", {
     tibble(a = c(1, 3), b = c(2, 4))
   )
 })
+
+test_that("Error if read_options$column_names and schema-names differ (ARROW-14744)", {
+
+  dst_dir  <- make_temp_dir()
+  dst_file <- file.path(dst_dir, "file.csv")
+  df <- df1[, c("int", "dbl")]
+  write.csv(df, dst_file, row.names = FALSE, quote = FALSE)
+
+  # Mismatch of column names vs schema given via read_options should raise an error
+  opts <- CsvReadOptions$create(column_names = c("i", "d"))
+  schema             <- schema(int = int32(), dbl = float64())
+
+  expect_error(
+    open_dataset(csv_dir, format = "csv", schema = schema, read_options = opts),
+    "column_names not matching or not found in schema-names"
+  )
```

Comment:
Thanks for the test, looking good.  One minor suggestion: how about swapping this for a simpler case, e.g.:
```
open_dataset(csv_dir, format = "csv", schema = schema, column_names = c("these", "are", "not", "the", "columns", "from", "the", "schema"))
```
It then removes the need to do the more complex work of `opts <- CsvReadOptions$create(column_names = c("i", "d"))` which is technically off-label usage that we don't recommend anyway.

Reply:
Thanks again for all of the comments, really helpful!

At the moment the example you provided would *not* raise an error if I've got it right. This is due to the following logic in `csv_file_format_read_opts`:

https://github.com/apache/arrow/blob/0fa4b9ca1e7b13ac230f85075d86d783eeea6a74/r/R/dataset-format.R#L223-L225

If schema is set, it overrides `column_names` in all situations. Because the check is done after `csv_file_format_read_opts` no error is raised despite of conflicting arguments from the user (which might be a little confusing for the user?).

Reply:
Good catch - let's update that to something like `if (!is.null(schema)) && is.null(opts[["column_names"]])`.